src/wiringx.c

```diff
@@ -103,6 +86,142 @@ static struct spi_t spi[2] = {
 	} while (0)
 #endif
 
+#ifndef __FreeBSD__
+/* SPI Bus Parameters */
+struct spi_t {
+	uint8_t mode;
+	uint8_t bits_per_word;
+	uint16_t delay;
+	uint32_t speed;
+	int fd;
+} spi_t;
+
+/* SPI states list implementation */
+
+struct spi_id_t {
+	uint8_t device;
+	uint8_t channel;
+};
+
+struct spi_list_t {
+	struct spi_id_t id;
+	struct spi_t data;
+};
+
+struct spi_list_t *spilist = 0;
+int nrspilist = 0;
+
+int spi_new(struct spi_id_t id) {
+	int entry = -1;
+	void *tmp = NULL;
+	int i = 0;
+
+	// loop over existing states to detect potential duplicates
+	for(i = 0; i < nrspilist; i++) {
+		if(memcmp(&spilist[i].id, &id, sizeof(struct spi_id_t)) == 0) {
+			entry = i;
+
+			/* clear existing entry */
+			if(spilist[entry].data.fd > 0) {
+				close(spilist[entry].data.fd);
+			}
+
+			/* return index of existing entry */
+			return entry;
+		}
+	}
+
+	/* reserve space for new entry (+1) */
+	entry = nrspilist++;
+	tmp = realloc(spilist, sizeof(struct spi_list_t)*nrspilist);
+	if(tmp == NULL) {
+		fprintf(stderr, "out of memory\n");
+		exit(EXIT_FAILURE);
+	}
+	spilist = tmp;
+	spilist[entry].id = id;
+	memset(&spilist[entry].data, 0, sizeof(struct spi_t));
+
+	/* return index of new entry */
+	return entry;
+}
+
+void spi_clear() {
+	struct spi_list_t *entry = NULL;
+	int i = 0;
+
+	/* loop over all entries */
+	for(i = 0; i < nrspilist; i++) {
+		entry = &spilist[i];
+
+		/* clear */
+		if(entry->data.fd > 0) {
+			close(entry->data.fd);
+		}
+	}
+
+	/* free storage */
+	free(spilist);
+	spilist = NULL;
+	nrspilist = 0;
+}
+
+/* test-case for list */
+int spi_test() {
+	int i = 0, j = 0, k = 0;
+	struct spi_id_t id;
+
+	/* try to grow the list a few times */
+	for(id.device = 0; id.device < 12; id.device++) {
+		for(id.channel = 0; id.channel < 12; id.channel++) {
+			k = spi_new(id);
+
+			/* 
+			 * check if the handle is unique
+			 * simplified method: predetermined handle
+			 */
+			if(k != 12*id.device + id.channel) {
+				wiringXLog(LOG_ERR, "Error: expected handle to spidev (%u, %u) = %i, but got %i.\n", id.device, id.channel, k);
```

Comment:
No need to insert `Error: ` in the log statements. The prefix should be handled by the log function.